token-lending/program/src/processor.rs

```diff
@@ -578,6 +587,315 @@ fn _deposit_reserve_liquidity<'a>(
+    if &reserve.lending_market != lending_market_info.key {
+        msg!("Reserve lending market does not match the lending market provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if &reserve.liquidity.supply_pubkey != source_liquidity_info.key {
+        msg!("Borrow reserve liquidity supply must be used as the source liquidity provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if &reserve.liquidity.supply_pubkey == destination_liquidity_info.key {
+        msg!(
+            "Borrow reserve liquidity supply cannot be used as the destination liquidity provided"
+        );
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if reserve.borrowing {
+        msg!("Multiple flash borrows not allowed");
+        return Err(LendingError::MultipleFlashBorrows.into());
+    }
+
+    let authority_signer_seeds = &[
+        lending_market_info.key.as_ref(),
+        &[lending_market.bump_seed],
+    ];
+    let lending_market_authority_pubkey =
+        Pubkey::create_program_address(authority_signer_seeds, program_id)?;
+    if &lending_market_authority_pubkey != lending_market_authority_info.key {
+        msg!(
+            "Derived lending market authority {} does not match the lending market authority provided {}",
+            &lending_market_authority_pubkey.to_string(),
+            &lending_market_authority_info.key.to_string(),
+        );
+        return Err(LendingError::InvalidMarketAuthority.into());
+    }
+
+    if reserve.config.fees.flash_loan_fee_wad == u64::MAX {
+        msg!("Flash loans are disabled for this reserve");
+        return Err(LendingError::FlashLoansDisabled.into());
+    }
+
+    let flash_loan_amount = liquidity_amount;
+    let flash_loan_amount_decimal = Decimal::from(flash_loan_amount);
+
+    // Make sure this isnt a cpi call
+    let current_index = load_current_index_checked(sysvar_info)? as usize;
+    let current_ixn = load_instruction_at_checked(current_index, sysvar_info)?;
+    if current_ixn.program_id != *program_id {
+        msg!(
+            "Cpi call found: Flash borrow from {} does not match program_id {} ",
+            current_ixn.program_id,
+            *program_id
+        );
+        return Err(LendingError::FlashBorrowCpi.into());
+    }
+
+    static FLASH_REPAY_INSTRUCTION: u8 = 18;
+
+    // loop through instructions, looking for an equivalent repay to this borrow
+    let mut i = current_index + 1;
+    loop {
+        // get the next instruction, die if theres no more
+        if let Ok(ixn) = load_instruction_at_checked(i, sysvar_info) {
+            // check if we have a toplevel repay toward the same reserve
+            // if so, confirm the amount, otherwise next instruction
+            // Note: First byte is instruction ID, next 8 bytes are liquidity_amount
+            if ixn.program_id == *program_id
+                && u8::from_le_bytes([ixn.data[0]]) == FLASH_REPAY_INSTRUCTION
```

Comment:
For the logic of parsing the flash repay data, is there a way we can parse the instruction data here using the same code path that the actual flash repay instruction uses?
Would be a nice to not have the instruction data parsing code in a bunch of different places